### PR TITLE
Use exponential backoff for connections / communications

### LINKED ISSUE #27: Improve reliability of gRPC connections / communications
**Motivation**

I've configured Falco Exporter as a sidecar, which exposes some certain issues, particularly the delay during Falco's BPF probe compilation and gRPC server binding. This causes the exporter to generally go into CrashLoopBackOff due to error.

**Feature**

Implement backoff retry functionality for gRPC client creation, connection and subscription.

```
2020/05/19 00:25:50 exporter error: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp [::1]:5060: connect: connection refused", retrying after backoff of 1 second(s)
2020/05/19 00:25:51 exporter error: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp [::1]:5060: connect: connection refused", retrying after backoff of 2 second(s)
2020/05/19 00:25:53 exporter error: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp [::1]:5060: connect: connection refused", retrying after backoff of 4 second(s)
2020/05/19 00:25:57 exporter error: rpc error: code = Unavailable desc = all SubConns are in TransientFailure, latest connection error: connection error: desc = "transport: Error while dialing dial tcp [::1]:5060: connect: connection refused", retrying after backoff of 8 second(s)
```

```
NAME          READY   STATUS    RESTARTS   AGE
falco-6ccmq   2/2     Running   0          58s
falco-hd2sh   2/2     Running   0          49s
falco-s74c5   2/2     Running   0          35s
falco-vm62v   2/2     Running   0          73s
falco-zhtc2   2/2     Running   0          82s
```

**Alternatives**

If I had used the current standard approach of multiple daemonsets/services I could have an initContainer waiting for the service (representing the Falco daemonset) to become available, with a sidecar it is not possible - I'd essentially need an initContainer _only_ for falco-exporter.

### PR DESCRIPTION
**Pull request by fujin**
**What type of PR is this?**

/kind feature

/area commands

**What this PR does / why we need it**:
Basic backoff functionality for client creation, connection and subscription.

**Which issue(s) this PR fixes**:

Fixes #27 

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
feat: implement exponential backoff algorithm on gRPC connection and subscription
```

### CODE REVIEW

**Comment by leogr** [comment]
Thank you for your contribution! 

SGTM, I will give it a try soon!

File: cmd/falco-exporter/root.go [hunk 1]
```diff
@@ -27,21 +30,46 @@ func main() {
...
+	var outputClient output.ServiceClient
+	for {
+		outputClient, err = c.Output()
+		if err != nil {
+			log.Printf("unable to obtain an output client: %v, retrying after backoff of %d second(s)", err, exp.NextDuration/time.Second)
+			exp.Backoff()
+		} else {
+			exp.Reset()
+			break
+		}
```
Comment (leogr) [inline cmd/falco-exporter/root.go:61]: [func Output() of client-go](https://github.com/falcosecurity/client-go/blob/3cb755d7594dc871f411b33b58a9366887bb4382/pkg/client/client.go#L75) checks for the connection then [just instantiates an object](https://github.com/falcosecurity/client-go/blob/3cb755d7594dc871f411b33b58a9366887bb4382/pkg/client/client.go#L80).

So I don't see how calling `c.Output()` multiple times could change the outcome, thus exponential backoff should not be used here.

Is there any other reason why we need exponential backoff here? If not, I would remove that.

File: cmd/falco-exporter/root.go [hunk 2]
```diff
@@ -27,21 +30,46 @@ func main() {
...
-	c, err := client.NewForConfig(config)
-	if err != nil {
-		log.Fatalf("unable to connect: %v", err)
+	var (
+		c   *client.Client
+		err error
+	)
+	exp := backoff.NewExponential(1*time.Second, 2*time.Minute)
+
+	// XXX(fujin): refactor into backoff() async func/closure or similar
```
Comment (leogr) [inline cmd/falco-exporter/root.go:39]:
```suggestion
	// TODO(fujin): refactor into backoff() async func/closure or similar
```

File: cmd/falco-exporter/root.go (hunk 1 above)
Comment (fujin) [reply cmd/falco-exporter/root.go:61]: Sounds good, will remove.

**Comment by fujin** [comment]
I encountered an issue overnight where this ran but got stuck trying to `back-off for 0 second(s)`; I'm going to reproduce and isolate/fix that case, delete the extra back-off, and comment.

**Comment by leogr** [comment]
Hi @fujin 

I made a new PR ( #29 ). Once we are sure it solved your issue I will close this.
Let me know!

**Comment by fujin** [comment]
closing in favor of #29 - tested OK in my env

**Pull request closed without merging**